ScriptureRenderingPipeline/Renderers/BibleRenderer.cs

```diff
@@ -16,152 +16,177 @@
+						ByteCount = byteCount
+					});
+				}
+				index.Bible.Add(outputBook);
+				downloadIndex.Content.Add(bookWithContent);
+				// Add whole.json for each chapter for book level fetching
+				outputTasks.Add(File.WriteAllTextAsync(Path.Join(destinationDir, abbreviation, "whole.json"), JsonSerializer.Serialize(bookWithContent)));
 
-            await Task.WhenAll(outputTasks);
-        }
-        /// <summary>
-        /// Load all USFM files in a directory inside of the ZipFileSystem
-        /// </summary>
-        /// <param name="directory">A ZipFileSystem to load from</param>
-        /// <returns>A list of USFM files</returns>
-        static async Task<List<USFMDocument>> LoadDirectoryAsync(ZipFileSystem directory)
-        {
-            USFMParser parser = new USFMParser(new List<string> { "s5" }, true);
-            var output = new List<USFMDocument>();
-            foreach (var f in directory.GetAllFiles(".usfm"))
-            {
-                var tmp = parser.ParseFromString(await directory.ReadAllTextAsync(f));
-                // If we don't have an abbreviation then try to figure it out from the file name
-                var tableOfContentsMarkers = tmp.GetChildMarkers<TOC3Marker>();
-                if(tableOfContentsMarkers.Count == 0)
-                {
-                    var bookAbbrivation = GetBookAbberviationFromFileName(f);
-                    if (bookAbbrivation != null)
-                    {
-                        tmp.Insert(new TOC3Marker() { BookAbbreviation = bookAbbrivation });
-                    }
-                }
-                else if (Utils.GetBookNumber(tableOfContentsMarkers[0].BookAbbreviation) == 0)
-                {
-                    var bookAbbrivation = GetBookAbberviationFromFileName(f);
-                    if (bookAbbrivation != null)
-                    {
-                        tableOfContentsMarkers[0].BookAbbreviation = bookAbbrivation;
-                    }
-                }
-                output.Add(tmp);
-            }
-            return output;
-        }
 
-        private static string GetBookAbberviationFromFileName(string f)
-        {
-            string bookAbbriviation = null;
-            var fileNameSplit = Path.GetFileNameWithoutExtension(f).Split('-');
-            if (fileNameSplit.Length == 2)
-            {
-                if (Utils.BibleBookOrder.Contains(fileNameSplit[1].ToUpper()))
-                {
-                    bookAbbriviation = fileNameSplit[1].ToUpper();
-                }
-            }
-            else if (fileNameSplit.Length == 1)
-            {
-                if (Utils.BibleBookOrder.Contains(fileNameSplit[0].ToUpper()))
-                {
-                    bookAbbriviation = fileNameSplit[0].ToUpper();
-                }
-            }
+				// Since the print all page isn't going to broken up then just write stuff out here
+				printBuilder.AppendLine(content);
+			}
+			long totalByteCount = downloadIndex.Content

```

Comment:
Linq is powerful but it can end up being a performance problem if you aren't careful